Doesn't build on hyprland .54

Using `hyprpm update` could not build with updated hyprland v0.54.0

```
❯ hyprpm add https://github.com/raybbian/hyprtasking.git -v
[v] parsed commit 0002f148c9a4fe421a9d33c0faa5528cdc411e62 at branch v0.54.0 on Fri Feb 27 18:03:19 2026, commits 6935, nix: false
[sudo] password for mknap:
[v] parsed commit 0002f148c9a4fe421a9d33c0faa5528cdc411e62 at branch v0.54.0 on Fri Feb 27 18:03:19 2026, commits 6935, nix: false

✔ Headers up to date.
! Disclaimer:
plugins, especially not official, have no guarantee of stability, availability or security.
Run them at your own risk.
This message will not appear again.
✔ adding a new plugin repository from https://github.com/raybbian/hyprtasking.git
  MAKE SURE that you trust the authors. DO NOT install random plugins without verifying the code and author.
  Are you sure? [Y/n]

→ Cloning https://github.com/raybbian/hyprtasking.git
✔ cloned
✔ found hyprpm manifest
✔ parsed manifest, found 1 plugins:
→ hyprtasking by raybbian version
→ Manifest has 15 pins, checking
✔ Hyprland headers OK
→ Building hyprtasking
 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╍━━━━━━━━━━━━━━━━━━━  3 / 5  Building plugin(s)[v] shell returned:  -> cd /run/user/1000/hyprpm/mknap && PKG_CONFIG_PATH="/var/cache/hyprpm/mknap/headersRoot/share/pkgconfig:$PKG_CONFIG_PATH" meson setup build --buildtype=debug
The Meson build system
Version: 1.10.1
Source dir: /run/user/1000/hyprpm/mknap
Build dir: /run/user/1000/hyprpm/mknap/build
Build type: native build
Project name: hyprtasking
Project version: 0.1
C++ compiler for the host machine: c++ (gcc 15.2.1 "c++ (GCC) 15.2.1 20260209")
C++ linker for the host machine: c++ ld.bfd 2.46
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C++ supports arguments -std=c++23: YES
meson.build:9: WARNING: Consider using the built-in option for language standard version instead of using "-std=c++23".
Found pkg-config: YES (/usr/bin/pkg-config) 2.5.1
Run-time dependency pixman-1 found: YES 0.46.4
Run-time dependency libdrm found: YES 2.4.131
Run-time dependency pangocairo found: YES 1.57.0
Run-time dependency libinput found: YES 1.30.1
Run-time dependency libudev found: YES 259
Run-time dependency wayland-server found: YES 1.24.0
Run-time dependency xkbcommon found: YES 1.13.1
Run-time dependency hyprland found: YES 0.54.0
Build targets in project: 1

hyprtasking 0.1

  User defined options
    buildtype: debug

Found ninja-1.13.2 at /usr/bin/ninja

 -> cd /run/user/1000/hyprpm/mknap && PKG_CONFIG_PATH="/var/cache/hyprpm/mknap/headersRoot/share/pkgconfig:$PKG_CONFIG_PATH" cd build && meson compile
[1/10] Compiling C++ object libhyprtasking.so.p/src_layout_linear.cpp.o
FAILED: [code=1] libhyprtasking.so.p/src_layout_linear.cpp.o
c++ -Ilibhyprtasking.so.p -I. -I.. -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/sysprof-6 -I/var/cache/hyprpm/mknap/headersRoot/include -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland/protocols -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -Wno-narrowing -shared -fPIC --no-gnu-unique -std=c++23 -fPIC -DWITH_GZFILEOP -pthread -MD -MQ libhyprtasking.so.p/src_layout_linear.cpp.o -MF libhyprtasking.so.p/src_layout_linear.cpp.o.d -o libhyprtasking.so.p/src_layout_linear.cpp.o -c ../src/layout/linear.cpp
../src/layout/linear.cpp:9:10: fatal error: hyprland/src/managers/LayoutManager.hpp: No such file or directory
    9 | #include <hyprland/src/managers/LayoutManager.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[2/10] Compiling C++ object libhyprtasking.so.p/src_input.cpp.o
FAILED: [code=1] libhyprtasking.so.p/src_input.cpp.o
c++ -Ilibhyprtasking.so.p -I. -I.. -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/sysprof-6 -I/var/cache/hyprpm/mknap/headersRoot/include -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland/protocols -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -Wno-narrowing -shared -fPIC --no-gnu-unique -std=c++23 -fPIC -DWITH_GZFILEOP -pthread -MD -MQ libhyprtasking.so.p/src_input.cpp.o -MF libhyprtasking.so.p/src_input.cpp.o.d -o libhyprtasking.so.p/src_input.cpp.o -c ../src/input.cpp
../src/input.cpp:6:10: fatal error: hyprland/src/managers/LayoutManager.hpp: No such file or directory
    6 | #include <hyprland/src/managers/LayoutManager.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[3/10] Compiling C++ object libhyprtasking.so.p/src_main.cpp.o
FAILED: [code=1] libhyprtasking.so.p/src_main.cpp.o
c++ -Ilibhyprtasking.so.p -I. -I.. -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/sysprof-6 -I/var/cache/hyprpm/mknap/headersRoot/include -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland/protocols -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -Wno-narrowing -shared -fPIC --no-gnu-unique -std=c++23 -fPIC -DWITH_GZFILEOP -pthread -MD -MQ libhyprtasking.so.p/src_main.cpp.o -MF libhyprtasking.so.p/src_main.cpp.o.d -o libhyprtasking.so.p/src_main.cpp.o -c ../src/main.cpp
../src/main.cpp:11:10: fatal error: hyprland/src/managers/LayoutManager.hpp: No such file or directory
   11 | #include <hyprland/src/managers/LayoutManager.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[4/10] Compiling C++ object libhyprtasking.so.p/src_pass_pass_element.cpp.o
[5/10] Compiling C++ object libhyprtasking.so.p/src_overview.cpp.o
FAILED: [code=1] libhyprtasking.so.p/src_overview.cpp.o
c++ -Ilibhyprtasking.so.p -I. -I.. -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/sysprof-6 -I/var/cache/hyprpm/mknap/headersRoot/include -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland/protocols -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -Wno-narrowing -shared -fPIC --no-gnu-unique -std=c++23 -fPIC -DWITH_GZFILEOP -pthread -MD -MQ libhyprtasking.so.p/src_overview.cpp.o -MF libhyprtasking.so.p/src_overview.cpp.o.d -o libhyprtasking.so.p/src_overview.cpp.o -c ../src/overview.cpp
../src/overview.cpp: In member function ‘void HTView::move_id(WORKSPACEID, bool)’:
../src/overview.cpp:159:47: error: no matching function for call to ‘Desktop::CFocusState::fullWindowFocus(PHLWINDOW&)’
  159 |         Desktop::focusState()->fullWindowFocus(hovered_window);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
../src/overview.cpp:159:47: note: there is 1 candidate
In file included from ../src/globals.hpp:6,
                 from ../src/config.hpp:6,
                 from ../src/overview.cpp:13:
/var/cache/hyprpm/mknap/headersRoot/include/hyprland/src/desktop/state/FocusState.hpp:31:32: note: candidate 1: ‘void Desktop::CFocusState::fullWindowFocus(PHLWINDOW, Desktop::eFocusReason, SP<CWLSurfaceResource>, bool)’
   31 |         void                   fullWindowFocus(PHLWINDOW w, eFocusReason reason, SP<CWLSurfaceResource> surface = nullptr, bool forceFSCycle = false);
      |                                ^~~~~~~~~~~~~~~
/var/cache/hyprpm/mknap/headersRoot/include/hyprland/src/desktop/state/FocusState.hpp:31:32: note: candidate expects 4 arguments, 1 provided
[6/10] Compiling C++ object libhyprtasking.so.p/src_manager.cpp.o
[7/10] Compiling C++ object libhyprtasking.so.p/src_render.cpp.o
[8/10] Compiling C++ object libhyprtasking.so.p/src_layout_grid.cpp.o
FAILED: [code=1] libhyprtasking.so.p/src_layout_grid.cpp.o
c++ -Ilibhyprtasking.so.p -I. -I.. -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/sysprof-6 -I/var/cache/hyprpm/mknap/headersRoot/include -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland/protocols -I/var/cache/hyprpm/mknap/headersRoot/include/hyprland -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -Wno-narrowing -shared -fPIC --no-gnu-unique -std=c++23 -fPIC -DWITH_GZFILEOP -pthread -MD -MQ libhyprtasking.so.p/src_layout_grid.cpp.o -MF libhyprtasking.so.p/src_layout_grid.cpp.o.d -o libhyprtasking.so.p/src_layout_grid.cpp.o -c ../src/layout/grid.cpp
../src/layout/grid.cpp: In member function ‘virtual bool HTLayoutGrid::should_render_window(PHLWINDOW)’:
../src/layout/grid.cpp:188:36: error: ‘class CInputManager’ has no member named ‘m_currentlyDraggedWindow’
  188 |     if (window == g_pInputManager->m_currentlyDraggedWindow.lock())
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~~
../src/layout/grid.cpp: In member function ‘virtual void HTLayoutGrid::render()’:
../src/layout/grid.cpp:454:55: error: ‘class CInputManager’ has no member named ‘m_currentlyDraggedWindow’
  454 |     const PHLWINDOW dragged_window = g_pInputManager->m_currentlyDraggedWindow.lock();
      |                                                       ^~~~~~~~~~~~~~~~~~~~~~~~
[9/10] Compiling C++ object libhyprtasking.so.p/src_layout_layout_base.cpp.o
ninja: build stopped: subcommand failed.
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja


✖ Plugin hyprtasking failed to build.
```

--- a/README.md
+++ b/README.md
@@ -4,7 +4,7 @@
 </div>
 
 > [!Important]
-> - Supports Hyprland release `v0.46.2-v0.52.1`.
+> - Supports Hyprland release `v0.46.2-v0.54.0`.
 
 https://github.com/user-attachments/assets/8d6cdfd2-2b17-4240-a117-1dbd2231ed4e
 

--- a/hyprpm.toml
+++ b/hyprpm.toml
@@ -2,7 +2,9 @@
 name = "hyprtasking"
 authors = ["raybbian"]
 commit_pins = [
-    # ["ab1d80f3d6aebd57a0971b53a1993b1c1dfe0b09", "40e6f615c6e2f7f016d1ee6cfa34668edbeb0cc1"], #v0.53.1
+    ["dd220efe7b1e292415bd0ea7161f63df9c95bfd3", "40e6f615c6e2f7f016d1ee6cfa34668edbeb0cc1"], #v0.53.3
+    ["39f3feddbee4a66be9608ed1eb7e73878d596b50", "40e6f615c6e2f7f016d1ee6cfa34668edbeb0cc1"], #v0.53.2
+    ["ab1d80f3d6aebd57a0971b53a1993b1c1dfe0b09", "40e6f615c6e2f7f016d1ee6cfa34668edbeb0cc1"], #v0.53.1
     ["ea444c35bb23b6e34505ab6753e069de7801cc25", "40e6f615c6e2f7f016d1ee6cfa34668edbeb0cc1"], #v0.53.0
     ["386376400119dd46a767c9f8c8791fd22c7b6e61", "9611bbd0db23bba9508da44f65989a7dc664d0a9"], #v0.52.2
     ["967c3c7404d4fa00234e29c70df3e263386d2597", "9611bbd0db23bba9508da44f65989a7dc664d0a9"], #v0.52.1

--- a/src/input.cpp
+++ b/src/input.cpp
@@ -3,7 +3,7 @@
 #include <hyprland/src/Compositor.hpp>
 #include <hyprland/src/macros.hpp>
 #include <hyprland/src/managers/KeybindManager.hpp>
-#include <hyprland/src/managers/LayoutManager.hpp>
+#include <hyprland/src/layout/LayoutManager.hpp>
 #include <hyprland/src/managers/PointerManager.hpp>
 #include <hyprland/src/managers/input/InputManager.hpp>
 
@@ -43,9 +43,13 @@ bool HTManager::start_window_drag() {
     g_pKeybindManager->changeMouseBindMode(MBIND_MOVE);
     g_pPointerManager->warpTo(mouse_coords);
 
-    const PHLWINDOW dragged_window = g_pInputManager->m_currentlyDraggedWindow.lock();
+    const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
+    if (target == nullptr)
+        return false;
+
+    const PHLWINDOW dragged_window = target->window();
     if (dragged_window != nullptr) {
-        if (dragged_window->m_draggingTiled) {
+        if (g_layoutManager->dragController()->draggingTiled()) {
             const Vector2D pre_pos = cursor_view->layout->local_ws_unscaled_to_global(
                 dragged_window->m_realPosition->value() - dragged_window->m_monitor->m_position,
                 workspace_id
@@ -92,10 +96,14 @@ bool HTManager::end_window_drag() {
         return false;
     }
 
+    const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
+    if (target == nullptr)
+        return false;
+
     // If not dragging window or drag is not move, then we just let go (supposed to prevent it
     // from messing up resize on border, but it should be good because above?)
-    const PHLWINDOW dragged_window = g_pInputManager->m_currentlyDraggedWindow.lock();
-    if (dragged_window == nullptr || g_pInputManager->m_dragMode != MBIND_MOVE) {
+    const PHLWINDOW dragged_window = target->window();
+    if (dragged_window == nullptr || g_layoutManager->dragController()->mode() != MBIND_MOVE) {
         g_pKeybindManager->changeMouseBindMode(MBIND_INVALID);
         return false;
     }

--- a/src/layout/grid.cpp
+++ b/src/layout/grid.cpp
@@ -9,6 +9,7 @@
 #include <hyprland/src/managers/animation/AnimationManager.hpp>
 #include <hyprland/src/managers/animation/DesktopAnimationManager.hpp>
 #include <hyprland/src/managers/input/InputManager.hpp>
+#include <hyprland/src/layout/LayoutManager.hpp>
 #include <hyprland/src/render/OpenGL.hpp>
 #include <hyprland/src/render/Renderer.hpp>
 #include <hyprland/src/render/pass/BorderPassElement.hpp>
@@ -21,6 +22,7 @@
 #include "../overview.hpp"
 #include "../render.hpp"
 #include "../types.hpp"
+#include "src/layout/target/Target.hpp"
 
 using Hyprutils::Utils::CScopeGuard;
 
@@ -184,8 +186,8 @@ bool HTLayoutGrid::should_render_window(PHLWINDOW window) {
     const PHLMONITOR monitor = get_monitor();
     if (window == nullptr || monitor == nullptr)
         return ori_result;
-
-    if (window == g_pInputManager->m_currentlyDraggedWindow.lock())
+    const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
+    if (target != nullptr && window == target->window())
         return false;
 
     PHLWORKSPACE workspace = window->m_workspace;
@@ -311,8 +313,8 @@ void HTLayoutGrid::render() {
 
     const float BORDERSIZE = HTConfig::value<Hyprlang::FLOAT>("border_size");
 
-    timespec time;
-    clock_gettime(CLOCK_MONOTONIC, &time);
+    const auto time = Time::steadyNow();
+
 
     g_pHyprRenderer->damageMonitor(monitor);
     g_pHyprOpenGL->m_renderData.pCurrentMonData->blurFBShouldRender = true;
@@ -384,7 +386,7 @@ void HTLayoutGrid::render() {
                 g_pHyprRenderer.get(),
                 monitor,
                 workspace,
-                &time,
+                time,
                 render_box
             );
 
@@ -401,7 +403,7 @@ void HTLayoutGrid::render() {
                 g_pHyprRenderer.get(),
                 monitor,
                 workspace,
-                &time,
+                time,
                 render_box
             );
         }
@@ -442,7 +444,7 @@ void HTLayoutGrid::render() {
                 g_pHyprRenderer.get(),
                 monitor,
                 start_workspace,
-                &time,
+                time,
                 render_box
             );
         }
@@ -451,7 +453,10 @@ void HTLayoutGrid::render() {
     const PHTVIEW cursor_view = ht_manager->get_view_from_cursor();
     if (cursor_view == nullptr)
         return;
-    const PHLWINDOW dragged_window = g_pInputManager->m_currentlyDraggedWindow.lock();
+    const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
+    if (target == nullptr)
+        return;
+    const PHLWINDOW dragged_window = target->window();
     if (dragged_window == nullptr)
         return;
     const Vector2D mouse_coords = g_pInputManager->getMouseCoordsInternal();
@@ -460,5 +465,5 @@ void HTLayoutGrid::render() {
                                 .scale(cursor_view->layout->drag_window_scale())
                                 .translate(mouse_coords);
     if (!window_box.intersection(monitor->logicalBox()).empty())
-        render_window_at_box(dragged_window, monitor, &time, window_box);
+        render_window_at_box(dragged_window, monitor, time, window_box);
 }

--- a/src/layout/grid.cpp
+++ b/src/layout/grid.cpp
@@ -1,6 +1,5 @@
 #include "grid.hpp"
 
-#include <ctime>
 #include <hyprland/src/Compositor.hpp>
 #include <hyprland/src/config/ConfigManager.hpp>
 #include <hyprland/src/config/ConfigValue.hpp>
@@ -186,6 +185,7 @@ bool HTLayoutGrid::should_render_window(PHLWINDOW window) {
     const PHLMONITOR monitor = get_monitor();
     if (window == nullptr || monitor == nullptr)
         return ori_result;
+
     const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
     if (target != nullptr && window == target->window())
         return false;

--- a/src/layout/linear.cpp
+++ b/src/layout/linear.cpp
@@ -6,7 +6,7 @@
 #include <hyprland/src/helpers/MiscFunctions.hpp>
 #include <hyprland/src/managers/animation/AnimationManager.hpp>
 #include <hyprland/src/managers/animation/DesktopAnimationManager.hpp>
-#include <hyprland/src/managers/LayoutManager.hpp>
+#include <hyprland/src/layout/LayoutManager.hpp>
 #include <hyprland/src/managers/input/InputManager.hpp>
 #include <hyprland/src/render/pass/BorderPassElement.hpp>
 #include <hyprland/src/render/pass/RectPassElement.hpp>
@@ -192,7 +192,11 @@ bool HTLayoutLinear::should_render_window(PHLWINDOW window) {
     if (window == nullptr || monitor == nullptr)
         return ori_result;
 
-    if (window == g_pInputManager->m_currentlyDraggedWindow.lock())
+    const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
+    if (target == nullptr)
+        return false;
+    const PHLWINDOW dragged_window = target->window();
+    if (window == dragged_window)
         return false;
 
     if (rendering_standard_ws)
@@ -307,8 +311,8 @@ void HTLayoutLinear::render() {
     const float BORDERSIZE = HTConfig::value<Hyprlang::FLOAT>("border_size");
     const float HEIGHT = HTConfig::value<Hyprlang::FLOAT>("linear:height") * monitor->m_scale;
 
-    timespec time;
-    clock_gettime(CLOCK_MONOTONIC, &time);
+    const auto time = Time::steadyNow();
+
 
     g_pHyprRenderer->damageMonitor(monitor);
     g_pHyprOpenGL->m_renderData.pCurrentMonData->blurFBShouldRender = true;
@@ -343,7 +347,7 @@ void HTLayoutLinear::render() {
         g_pHyprRenderer.get(),
         monitor,
         big_ws,
-        &time,
+        time,
         mon_box
     );
 
@@ -417,7 +421,7 @@ void HTLayoutLinear::render() {
                 g_pHyprRenderer.get(),
                 monitor,
                 workspace,
-                &time,
+                time,
                 render_box
             );
 
@@ -434,7 +438,7 @@ void HTLayoutLinear::render() {
                 g_pHyprRenderer.get(),
                 monitor,
                 workspace,
-                &time,
+                time,
                 render_box
             );
         }
@@ -453,7 +457,12 @@ void HTLayoutLinear::render() {
     const PHTVIEW cursor_view = ht_manager->get_view_from_cursor();
     if (cursor_view == nullptr)
         return;
-    const PHLWINDOW dragged_window = g_pInputManager->m_currentlyDraggedWindow.lock();
+
+    const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
+    if (target == nullptr)
+        return;
+
+    const PHLWINDOW dragged_window = target->window();
     if (dragged_window == nullptr)
         return;
     const Vector2D mouse_coords = g_pInputManager->getMouseCoordsInternal();
@@ -462,5 +471,5 @@ void HTLayoutLinear::render() {
                                 .scale(cursor_view->layout->drag_window_scale())
                                 .translate(mouse_coords);
     if (!window_box.intersection(monitor->logicalBox()).empty())
-        render_window_at_box(dragged_window, monitor, &time, window_box);
+        render_window_at_box(dragged_window, monitor, time, window_box);
 }

--- a/src/layout/linear.cpp
+++ b/src/layout/linear.cpp
@@ -193,10 +193,7 @@ bool HTLayoutLinear::should_render_window(PHLWINDOW window) {
         return ori_result;
 
     const SP<Layout::ITarget> target = g_layoutManager->dragController()->target();
-    if (target == nullptr)
-        return false;
-    const PHLWINDOW dragged_window = target->window();
-    if (window == dragged_window)
+    if (target != nullptr && window == target->window())
         return false;
 
     if (rendering_standard_ws)

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -314,10 +314,20 @@ static void init_functions() {
     Log::logger->log(LOG, "[Hyprtasking] Attempting hook {}", FNS2[0].signature);
     success = should_render_window_hook->hook() && success;
 
-    static auto FNS3 = HyprlandAPI::findFunctionsByName(PHANDLE, "renderWindow");
+    // Right now (in v0.54.0) there are several "renderWindow" functions
+    // This is needed so it won't break on update that adds/removes a
+    // function with this name
+    // This, however, requires checking for signautre changes
+    static auto FNS3 = HyprlandAPI::findFunctionsByName(
+        PHANDLE,
+        "_ZN13CHyprRenderer12renderWindowEN9Hyprutils6Memory14CSha"
+        "redPointerIN7Desktop4View7CWindowEEENS2_I8CMonitorEERKNSt"
+        "6chrono10time_pointINS9_3_V212steady_clockENS9_8durationI"
+        "lSt5ratioILl1ELl1000000000EEEEEEb15eRenderPassModebb"
+    );
     if (FNS3.empty())
         fail_exit("No renderWindow");
-    render_window = FNS3[1].address;
+    render_window = FNS3[0].address;
 
     static auto FNS4 = HyprlandAPI::findFunctionsByName(PHANDLE, "isSolitaryBlocked");
     if (FNS4.empty())

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -1,6 +1,5 @@
 #include <linux/input-event-codes.h>
 
-#include <ctime>
 #include <hyprland/src/Compositor.hpp>
 #include <hyprland/src/SharedDefs.hpp>
 #include <hyprland/src/desktop/DesktopTypes.hpp>
@@ -19,7 +18,6 @@
 #include <hyprlang.hpp>
 #include <hyprutils/math/Box.hpp>
 #include <hyprutils/math/Vector2D.hpp>
-#include <print>
 
 #include "config.hpp"
 #include "globals.hpp"
@@ -91,17 +89,12 @@ static SDispatchResult dispatch_toggle_view(std::string arg) {
 }
 
 static SDispatchResult dispatch_move(std::string arg) {
-    std::print("dispatch move");
     if (ht_manager == nullptr)
         return {.success = false, .error = "ht_manager is null"};
-    std::print("ht_man is ok");
     const PHTVIEW cursor_view = ht_manager->get_view_from_cursor();
-    std::print("got view from cursor");
     if (cursor_view == nullptr)
         return {.success = false, .error = "cursor_view is null"};
-    std::print("view not null ptr, moving");
     cursor_view->move(arg, false);
-    std::print("moved, returning");
     return {};
 }
 

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -8,16 +8,18 @@
 #include <hyprland/src/helpers/Monitor.hpp>
 #include <hyprland/src/macros.hpp>
 #include <hyprland/src/managers/KeybindManager.hpp>
-#include <hyprland/src/managers/LayoutManager.hpp>
+#include <hyprland/src/layout/LayoutManager.hpp>
 #include <hyprland/src/managers/PointerManager.hpp>
 #include <hyprland/src/managers/input/InputManager.hpp>
 #include <hyprland/src/plugins/HookSystem.hpp>
 #include <hyprland/src/plugins/PluginAPI.hpp>
 #include <hyprland/src/plugins/PluginSystem.hpp>
 #include <hyprland/src/render/Renderer.hpp>
+#include <hyprland/src/event/EventBus.hpp>
 #include <hyprlang.hpp>
 #include <hyprutils/math/Box.hpp>
 #include <hyprutils/math/Vector2D.hpp>
+#include <print>
 
 #include "config.hpp"
 #include "globals.hpp"
@@ -89,12 +91,17 @@ static SDispatchResult dispatch_toggle_view(std::string arg) {
 }
 
 static SDispatchResult dispatch_move(std::string arg) {
+    std::print("dispatch move");
     if (ht_manager == nullptr)
         return {.success = false, .error = "ht_manager is null"};
+    std::print("ht_man is ok");
     const PHTVIEW cursor_view = ht_manager->get_view_from_cursor();
+    std::print("got view from cursor");
     if (cursor_view == nullptr)
         return {.success = false, .error = "cursor_view is null"};
+    std::print("view not null ptr, moving");
     cursor_view->move(arg, false);
+    std::print("moved, returning");
     return {};
 }
 
@@ -128,7 +135,7 @@ static void hook_render_workspace(
     void* thisptr,
     PHLMONITOR monitor,
     PHLWORKSPACE workspace,
-    timespec* now,
+   const Time::steady_tp& now,
     const CBox& geometry
 ) {
     if (ht_manager == nullptr) {
@@ -169,7 +176,7 @@ static uint32_t hook_is_solitary_blocked(void* thisptr, bool full) {
     return (*(origIsSolitaryBlocked)is_solitary_blocked_hook->m_original)(thisptr, full);
 }
 
-static void on_mouse_button(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_mouse_button(IPointer::SButtonEvent e, Event::SCallbackInfo& info) {
     if (ht_manager == nullptr)
         return;
 
@@ -177,7 +184,6 @@ static void on_mouse_button(void* thisptr, SCallbackInfo& info, std::any args) {
     if (cursor_view == nullptr)
         return;
 
-    const auto e = std::any_cast<IPointer::SButtonEvent>(args);
     const bool pressed = e.state == WL_POINTER_BUTTON_STATE_PRESSED;
 
     const unsigned int drag_button = HTConfig::value<Hyprlang::INT>("drag_button");
@@ -192,41 +198,37 @@ static void on_mouse_button(void* thisptr, SCallbackInfo& info, std::any args) {
     }
 }
 
-static void on_mouse_move(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_mouse_move(Vector2D c, Event::SCallbackInfo& info) {
     if (ht_manager == nullptr)
         return;
     info.cancelled = ht_manager->on_mouse_move();
 }
 
-static void on_mouse_axis(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_mouse_axis(IPointer::SAxisEvent e, Event::SCallbackInfo& info) {
     if (ht_manager == nullptr)
         return;
-    const auto e = std::any_cast<IPointer::SAxisEvent>(
-        std::any_cast<std::unordered_map<std::string, std::any>>(args)["event"]
-    );
     info.cancelled = ht_manager->on_mouse_axis(e.delta);
 }
 
-static void on_swipe_begin(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_swipe_begin(IPointer::SSwipeBeginEvent e, Event::SCallbackInfo& info) {
     if (ht_manager == nullptr)
         return;
     ht_manager->swipe_start();
 }
 
-static void on_swipe_update(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_swipe_update(IPointer::SSwipeUpdateEvent e, Event::SCallbackInfo& info) {
     if (ht_manager == nullptr)
         return;
-    auto e = std::any_cast<IPointer::SSwipeUpdateEvent>(args);
     info.cancelled = ht_manager->swipe_update(e);
 }
 
-static void on_swipe_end(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_swipe_end(IPointer::SSwipeEndEvent e, Event::SCallbackInfo& info) {
     if (ht_manager == nullptr)
         return;
     info.cancelled = ht_manager->swipe_end();
 }
 
-static void cancel_event(void* thisptr, SCallbackInfo& info, std::any args) {
+static void cancel_event(Event::SCallbackInfo& info) {
     if (ht_manager == nullptr || !ht_manager->cursor_view_active())
         return;
     info.cancelled = true;
@@ -276,7 +278,7 @@ static void register_monitors() {
     }
 }
 
-static void on_config_reloaded(void* thisptr, SCallbackInfo& info, std::any args) {
+static void on_config_reloaded() {
     notify_config_changes();
 
     if (ht_manager == nullptr)
@@ -326,7 +328,7 @@ static void init_functions() {
 
     static auto FNS4 = HyprlandAPI::findFunctionsByName(PHANDLE, "isSolitaryBlocked");
     if (FNS4.empty())
-        fail_exit("No isSolitaryBlocked!");
+        fail_exit("No isSolitaryBlocked");
 
     is_solitary_blocked_hook =
         HyprlandAPI::createFunctionHook(PHANDLE, FNS4[0].address, (void*)hook_is_solitary_blocked);
@@ -338,27 +340,23 @@ static void init_functions() {
 }
 
 static void register_callbacks() {
-    static auto P1 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "mouseButton", on_mouse_button);
-    static auto P2 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "mouseMove", on_mouse_move);
-    static auto P3 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "mouseAxis", on_mouse_axis);
+    static auto P1 = Event::bus()->m_events.input.mouse.button.listen(on_mouse_button);
+    static auto P2 = Event::bus()->m_events.input.mouse.move.listen(on_mouse_move);
+    static auto P3 = Event::bus()->m_events.input.mouse.axis.listen(on_mouse_axis);
 
     // TODO: support touch
-    static auto P4 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "touchDown", cancel_event);
-    static auto P5 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "touchUp", cancel_event);
-    static auto P6 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "touchMove", cancel_event);
+    static auto P4 = Event::bus()->m_events.input.touch.down.listen([&] (ITouch::SDownEvent e, Event::SCallbackInfo i) { cancel_event(i); } );
+    static auto P5 = Event::bus()->m_events.input.touch.up.listen([&] (ITouch::SUpEvent e, Event::SCallbackInfo i) { cancel_event(i); } );
+    static auto P6 = Event::bus()->m_events.input.touch.motion.listen([&] (ITouch::SMotionEvent e, Event::SCallbackInfo i) { cancel_event(i); } );
+    // static auto P7 = Event::bus()->m_events.input.touch.cancel.listen([&] (ITouch::SCancelEvent e, Event::SCallbackInfo i) { cancel_event(i); } );
 
-    static auto P7 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "swipeBegin", on_swipe_begin);
-    static auto P8 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "swipeUpdate", on_swipe_update);
-    static auto P9 = HyprlandAPI::registerCallbackDynamic(PHANDLE, "swipeEnd", on_swipe_end);
 
-    static auto P10 =
-        HyprlandAPI::registerCallbackDynamic(PHANDLE, "configReloaded", on_config_reloaded);
+    static auto P7 = Event::bus()->m_events.gesture.swipe.begin.listen(on_swipe_begin);
+    static auto P8 = Event::bus()->m_events.gesture.swipe.update.listen(on_swipe_update);
+    static auto P9 = Event::bus()->m_events.gesture.swipe.end.listen(on_swipe_end);
 
-    static auto P11 = HyprlandAPI::registerCallbackDynamic(
-        PHANDLE,
-        "monitorAdded",
-        [&](void* thisptr, SCallbackInfo& info, std::any data) { register_monitors(); }
-    );
+    static auto P10 = Event::bus()->m_events.config.reloaded.listen(on_config_reloaded);
+    static auto P11 = Event::bus()->m_events.monitor.added.listen(register_monitors);
 }
 
 static void add_dispatchers() {

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -317,7 +317,7 @@ static void init_functions() {
     static auto FNS3 = HyprlandAPI::findFunctionsByName(PHANDLE, "renderWindow");
     if (FNS3.empty())
         fail_exit("No renderWindow");
-    render_window = FNS3[0].address;
+    render_window = FNS3[1].address;
 
     static auto FNS4 = HyprlandAPI::findFunctionsByName(PHANDLE, "isSolitaryBlocked");
     if (FNS4.empty())

--- a/src/overview.cpp
+++ b/src/overview.cpp
@@ -157,7 +157,7 @@ void HTView::move_id(WORKSPACEID ws_id, bool move_window) {
 
     monitor->changeWorkspace(other_workspace);
     if (move_window) {
-        Desktop::focusState()->fullWindowFocus(hovered_window, Desktop::FOCUS_REASON_DESKTOP_STATE_CHANGE);
+        Desktop::focusState()->fullWindowFocus(hovered_window, Desktop::FOCUS_REASON_CLICK);
         warp = *CConfigValue<Hyprlang::INT>("plugin:hyprtasking:warp_on_move_window");
     } else {
         warp = *CConfigValue<Hyprlang::INT>("cursor:warp_on_change_workspace");

--- a/src/overview.cpp
+++ b/src/overview.cpp
@@ -14,6 +14,7 @@
 #include "globals.hpp"
 #include "layout/grid.hpp"
 #include "layout/linear.hpp"
+#include "src/desktop/state/FocusState.hpp"
 
 HTView::HTView(MONITORID in_monitor_id) {
     monitor_id = in_monitor_id;
@@ -156,7 +157,7 @@ void HTView::move_id(WORKSPACEID ws_id, bool move_window) {
 
     monitor->changeWorkspace(other_workspace);
     if (move_window) {
-        Desktop::focusState()->fullWindowFocus(hovered_window);
+        Desktop::focusState()->fullWindowFocus(hovered_window, Desktop::FOCUS_REASON_DESKTOP_STATE_CHANGE);
         warp = *CConfigValue<Hyprlang::INT>("plugin:hyprtasking:warp_on_move_window");
     } else {
         warp = *CConfigValue<Hyprlang::INT>("cursor:warp_on_change_workspace");

--- a/src/render.cpp
+++ b/src/render.cpp
@@ -35,6 +35,7 @@ void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, const Time::stea
 
     g_pHyprRenderer->damageWindow(window);
     ((render_window_t)render_window)(
+        // g_pHyprRenderer.get(),
         window,
         monitor,
         time,

--- a/src/render.cpp
+++ b/src/render.cpp
@@ -1,6 +1,5 @@
 #include "render.hpp"
 
-#include <ctime>
 #include <hyprland/src/Compositor.hpp>
 #include <hyprland/src/config/ConfigValue.hpp>
 #include <hyprland/src/desktop/DesktopTypes.hpp>
@@ -11,11 +10,12 @@
 #include <hyprutils/math/Vector2D.hpp>
 
 #include "globals.hpp"
+#include "src/helpers/time/Time.hpp"
 #include "types.hpp"
 
 // Note: box is relative to (0, 0), not monitor
-void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, timespec* time, CBox box) {
-    if (!window || !monitor || !time)
+void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, const Time::steady_tp& time, CBox box) {
+    if (!window || !monitor)
         return;
 
     box.x -= monitor->m_position.x;
@@ -35,7 +35,6 @@ void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, timespec* time, 
 
     g_pHyprRenderer->damageWindow(window);
     ((render_window_t)render_window)(
-        g_pHyprRenderer.get(),
         window,
         monitor,
         time,

--- a/src/render.cpp
+++ b/src/render.cpp
@@ -35,14 +35,14 @@ void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, const Time::stea
 
     g_pHyprRenderer->damageWindow(window);
     ((render_window_t)render_window)(
-        // g_pHyprRenderer.get(),
+        g_pHyprRenderer.get(),
         window,
         monitor,
         time,
         true,
         RENDER_PASS_MAIN,
         false,
-        false
+        true
     );
 
     g_pHyprRenderer->m_renderPass.add(makeUnique<CRendererHintsPassElement>(

--- a/src/render.hpp
+++ b/src/render.hpp
@@ -2,6 +2,7 @@
 
 #include <hyprland/src/desktop/DesktopTypes.hpp>
 #include <hyprland/src/plugins/PluginAPI.hpp>
+#include <hyprland/src/helpers/time/Time.hpp>
 #include <hyprutils/math/Box.hpp>
 
-void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, timespec* time, CBox box);
+void render_window_at_box(PHLWINDOW window, PHLMONITOR monitor, const Time::steady_tp& time, CBox box);

--- a/src/types.hpp
+++ b/src/types.hpp
@@ -4,26 +4,23 @@
 #include <hyprland/src/desktop/DesktopTypes.hpp>
 #include <hyprland/src/render/Renderer.hpp>
 #include <hyprutils/math/Box.hpp>
-
 typedef void (*render_workspace_t)(
     void* thisptr,
     PHLMONITOR pMonitor,
     PHLWORKSPACE pWorkspace,
-    timespec* now,
+    const Time::steady_tp& now,
     const CBox& geometry
 );
 
 typedef bool (*should_render_window_t)(void* thisptr, PHLWINDOW pWindow, PHLMONITOR pMonitor);
-
 typedef void (*render_window_t)(
-    void* thisptr,
     PHLWINDOW pWindow,
     PHLMONITOR pMonitor,
-    timespec* time,
+    const Time::steady_tp& time,
     bool decorate,
     eRenderPassMode mode,
     bool ignorePosition,
-    bool ignoreAllGeometry
+    bool standalone
 );
 
 typedef long VIEWID;

--- a/src/types.hpp
+++ b/src/types.hpp
@@ -14,6 +14,7 @@ typedef void (*render_workspace_t)(
 
 typedef bool (*should_render_window_t)(void* thisptr, PHLWINDOW pWindow, PHLMONITOR pMonitor);
 typedef void (*render_window_t)(
+    // void* thisptr,
     PHLWINDOW pWindow,
     PHLMONITOR pMonitor,
     const Time::steady_tp& time,

--- a/src/types.hpp
+++ b/src/types.hpp
@@ -14,7 +14,7 @@ typedef void (*render_workspace_t)(
 
 typedef bool (*should_render_window_t)(void* thisptr, PHLWINDOW pWindow, PHLMONITOR pMonitor);
 typedef void (*render_window_t)(
-    // void* thisptr,
+    void* thisptr,
     PHLWINDOW pWindow,
     PHLMONITOR pMonitor,
     const Time::steady_tp& time,

